Duke Chatbot by Yaladah

Yaladah opened the pull request:
# Duke chatbot
>Every day, my daddy told me the same thing. 'Once a task is just begun, never leave it till it's done. Be the labour great or small, do it well or not at all.' - Quincy Jones

Chatbot with functions:
- Add todo tasks
- Add deadline tasks with single deadline
- Add event tasks with a start time and end time
- Function to mark and unmark tasks as done or undone
- Function to delete unwanted tasks
- Function to list down recorded tasks
- Function to close chatbot

Steps to use the chatbot,
1. Download the jar file [here](https://github.com/Yaladah/ip/releases/tag/A-Jar)
2. Navigate to the directory containing the jar file.
3. Open the terminal and run the file using `java -jar duke.jar`
4. Have **FUN** with it!

What you can do,
- [x] Add tasks
- [x] Manage tasks
- [ ] Edit tasks (Coming soon)

The `main` method in the program looks like this,
```
public static void main(String[] args) {
        new Duke(SAVED_PATH).run();
}
```

Commit:
Add Gradle support

Yaladah committed:
feature: Added function for chatbot to echo the user inputs, and to exit the process when 'bye' is input.

Yaladah committed:
feature: Added function to add inputs to a list of tasks. When "list" is input, displays the list of currently input tasks.

Yaladah committed:
feature: Added function to mark and unmark tasks.
- Also included exception catch to prevent out of bound searches

Yaladah committed:
Removed commented out lines in Duke.java

Yaladah committed:
Feature: Added more types of tasks.
- Todo tasks have no time specification.
- Deadline tasks that are to be completed by a certain date.
- Event tasks are tasks with a time window to be completed by.

Yaladah committed:
Feature: Added text-ui testing script to test basic functions.

Yaladah committed:
Feature: Added error catching to existing functions to prevent incorrect inputs.
- Also updated the test script to include DukeExceptions caught

Yaladah committed:
Feature: Added funtion to delete tasks from task list.

Yaladah committed:
Feature: Add functionality to save tasks to harddrive

Yaladah committed:
Chore: Update automated testing scripts

Yaladah committed:
Feature: Add ability for chatbot to recognise dates
- Able to read input dates in yyyy-mm-dd format and display on list as dd-MMM-yyyy (eg 23 Feb 2023)

evanpy commented on src/main/java/Duke.java:
```diff
@@ -133,14 +150,15 @@ public boolean readInput(String input) throws DukeException, IOException {
-        return true;
     }
 
     public void load() {
```
Perhaps can include subject that is to be loaded. Otherwise, good code 👍🏻

evanpy reviewed (commented):
LGTM! 😄

Huggenguggen commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,240 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+    public static String projName = " ____        _        \n"
+                                    + "|  _ \\ _   _| | _____ \n"
+                                    + "| | | | | | | |/ / _ \\\n"
+                                    + "| |_| | |_| |   <  __/\n"
+                                    + "|____/ \\__,_|_|\\_\\___|\n";
```
Why not make this a final as well? 
i.e. 
```
public final static String PROJ_NAME = "..."
```

Huggenguggen commented on src/main/java/EventTask.java:
```diff
@@ -0,0 +1,25 @@
+    return "[E] " + super.toString() + " (from: " + this.startTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy"))
+            + " to: " + this.endTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy")) + ")";
+  }
+}
```
Very Clean and all to the coding standards! :)

Yaladah committed:
Chore: Abstract out closely related code as classes

Yaladah committed:
Chore: Categorise classes into more concrete packages

Yaladah committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Yaladah committed:
Chore: Add gradle support to automate testing

Yaladah committed:
Add Junit test cases

Yaladah committed:
Fix: Update dependencies to build Jar files

Yaladah committed:
Chore: Add JavaDocs to most non-trivial methods

Yaladah committed:
Chore: Touch up on code to comply to java coding standards

Yaladah committed:
Feature: Add function to filter list of tasks by keyword

Bentimate commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return String input by the user.
+     */
+    public String requestUserInput() {
```
Good use of verbs for your methods!

Bentimate commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,138 @@
+     * @throws DukeException Thrown when unexpected behaviour occurs with Duke.
+     */
+    public Task[] load() throws DukeException {
+        File f = new File(SAVED_PATH);
```
Should this be named "file" instead?

Bentimate reviewed (commented):
Overall very clean and well done!

Yaladah committed:
Feature: Implement the use of GUI using JavaFX

Yaladah committed:
Chore: Fix 'bye' function

Yaladah committed:
Chore: Update README

Yaladah committed:
Chore: Add assertions

Yaladah committed:
Chore: Touch up on code quality issues

Yaladah committed:
Merge pull request #3 from Yaladah/branch-A-CodeQuality

Chore: Touch up on code quality issues

Yaladah committed:
Merge pull request #2 from Yaladah/branch-A-Assertions

Chore: Add assertions

Yaladah committed:
Chore: Fix coding quality

Yaladah committed:
Feat: Add duration tasks
- Tasks that have a given time period to be completed by

Yaladah committed:
Chore: Update README and user guide

Yaladah committed:
Chore: update build.gradle information

Yaladah committed:
Fix save feature
- Instead of typing command 'save', tasklist is updated to hard drive everytime tasklist is changed

Yaladah committed:
Chore: Change variable name

Yaladah committed:
Chore: Fix Ui.png file name

Yaladah committed:
Chore: Remove unused code

Yaladah committed:
Chore: Reduce method lengths and standardise javadocs
